arm-service-map/2015-11-01-preview/swagger/arm-service-map.json

```diff
@@ -0,0 +1,2508 @@
+        }
+      },
+      "x-ms-external": true,
+      "x-ms-azure-resource": true,
+      "required": [
+        "id",
+        "type",
+        "name"
+      ]
+    },
+    "ResourceReference": {
+      "description": "Represents a reference to another resource.",
+      "discriminator": "kind",
+      "properties": {
+        "id": {
+          "type": "string",
+          "description": "Resource URI."
+        },
+        "type": {
+          "type": "string",
+          "description": "Additional resource type qualifier.  Always present when retrieving data from the service.  Optional when sending data to the service."
+        },
+        "name": {
+          "type": "string",
+          "description": "Resource name. Always present when retrieving data from the service.  Optional when sending data to the service."
+        },
+        "kind": {
+          "type": "string",
+          "description": "Specifies the sub-class of the reference."
+        }
+      },
+      "required": [
+        "id"
+      ],
+      "x-ms-azure-resource": true
+    },
+    "ResourceOnMachineReference": {
+      "description": "Reference to a resource on a machine.",
+      "allOf": [
+        {
+          "$ref": "#/definitions/ResourceReference"
+        }
+      ],
+      "x-ms-discriminator-value": "ref:onmachine",
+      "properties": {
+        "properties": {
+          "x-ms-client-flatten": true,
+          "type": "object",
+          "description": "Resource properties.",
+          "properties": {
+            "machine": {
+              "$ref": "#/definitions/ResourceReference"
+            }
+          },
+          "required": [
+            "machine"
+          ]
+        }
+      }
+    },
+    "CoreResource": {
+      "description": "Marker resource for the core Service Map resources",
+      "allOf": [
+        {
+          "$ref": "#/definitions/Resource"
+        }
+      ],
+      "discriminator": "kind",
+      "required": [
+        "kind"
+      ]
+    },
+    "Machine": {
+      "description": "A machine resource represents a discovered computer system. It can be *monitored*, i.e., a Dependency Agent is running on it, or *discovered*, i.e., its existence was inferred by observing the data stream from monitored machines. As machines change, prior versions of the machine resource are preserved and available for access. A machine is live during an interval of time, if either its Dependency Agent has reported data during (parts) of that interval, or a Dependency agent running on other machines has reported activity associated with the machine.",
+      "allOf": [
+        {
+          "$ref": "#/definitions/CoreResource"
+        }
+      ],
+      "x-ms-discriminator-value": "machine",
+      "properties": {
+        "properties": {
+          "x-ms-client-flatten": true,
+          "type": "object",
+          "description": "Resource properties.",
+          "properties": {
```

Comment:
Do not define objects inline. Please define them as a separate model definition and reference them. This educes confusion and enhances readability. 

Which of these model properties are required and which are readOnly?